File: coursera/py/largest_number.py
#Uses python3
#encoding: utf-8

""""
answer = empty string
while Digits is not empty:
    maxDigit = -00
    for digit in Digits:
        if digit ≥ maxDigit:
            maxDigit = digit

        append maxDigit to answer
        remove maxDigit from Digits
return answer
"""""

def is_greater_or_equal(a, b):
    if len(str(a)) != len(str(b)):
        res1 = "{}{}".format(a, b)
        res2 = "{}{}".format(b, a)
        response = int(res1) >= int(res2)
    else:
        response = a >= b
    return response


def largest_number(a):
    a.sort(reverse=True)
    res = ""
    while len(a) > 0:
        maxdigit = 0
        index = -1
        for i in range(len(a)):
            if is_greater_or_equal(a[i], maxdigit):
                maxdigit = a[i]
                index = i
        res += str(a[index])
        del a[index]
    return res


def test0():
    a = [1, 5, 4, 9]
    print(largest_number(a))

def test1():
    n = 2
    a = [21, 2]
    print(largest_number(a))
    output = 221

def test2():
    n = 5
    a = [9, 4, 6, 1, 9]
    print(largest_number(a))
    output = 99641

def test3():
    n = 3
    a = [23, 39, 92]
    print(largest_number(a))
    output = 923923

# test1()
# test2()
# test3()

if __name__ == '__main__':
    import sys
    n = int(input())
    a = [int(x) for x in input().split()]

    print(largest_number(a))
    

File: coursera/cpp/largest_number.cc
#include <iostream>
#include <vector>
// #include <iomanip>
#include <algorithm>
#include <string>

using namespace std;

bool comparator(int i, int j){
  string istr = to_string(i);
  string jstr = to_string(j);

  if(istr.size() != jstr.size()){
    string ij = istr + jstr;
    string ji = jstr + istr;
    return stoi(ij) >= stoi(ji);
  } else {
    return i >= j;
  }
}

string largest_number(vector<int> a){
  sort(a.begin(), a.end(), comparator);

  string res = "";
  for (size_t i = 0; i < a.size(); i++) {
    res += to_string(a[i]);
  }
  return res;
}

void test0(){
  std::vector<int> v;
  v = {21, 2};
  std::cout << "r: " << largest_number(v) << std::endl;
}

void test1(){
  std::vector<int> v;
  v = {9, 4, 6, 1, 9};
  std::cout << "r: " << largest_number(v) << std::endl;
}

void test2(){
  std::vector<int> v;
  v = {23, 39, 92};
  std::cout << "r: " << largest_number(v) << std::endl;
}

int main1(){
  // test0();
  // test1();
  test2();
  return 0;
}

int main(){
  int n;
  std::cin >> n;
  vector<int> values(n);
  for (int i = 0; i < n; i++) {
    std::cin >> values[i];
  }
 std::cout << largest_number(values) << std::endl;
  return 0;
}
